src/main/java/cinema/security/AuthenticationServiceImpl.java

```diff
@@ -0,0 +1,38 @@
+package cinema.security;
+
+import cinema.exceptions.AuthenticationException;
+import cinema.lib.Inject;
+import cinema.lib.Service;
+import cinema.model.User;
+import cinema.service.UserService;
+import cinema.util.HashUtil;
+import java.util.Optional;
+
+@Service
+public class AuthenticationServiceImpl implements AuthenticationService {
+
+    @Inject
+    private UserService userService;
+
+    @Override
+    public User login(String email, String password) throws AuthenticationException {
+        Optional<User> user = userService.findByEmail(email);
+        if (user.isPresent() && user.get().getPassword()
+                .equals(HashUtil.hashPassword(password, user.get().getSalt()))) {
+            return user.get();
+        }
+        throw new AuthenticationException("Wrong password or login");
+    }
+
+    @Override
+    public User register(String email, String password) {
+        User user = new User();
+        user.setEmail(email);
+        byte[] salt = HashUtil.getSalt();
+        String saltedPassword = HashUtil.hashPassword(password, salt);
+        user.setPassword(saltedPassword);
+        user.setSalt(salt);
```

Comment:
All the business logic somehow related to initializing or setting the values to our fields must be on the **_service_** layer.
If somebody creates the object **_User_** not throughout the register method but using the **_userService.create()_** instead, we'll not have its password hashed, which might cause some mistakes in the future while working with the object or with the security itself.
So let's move the hashing logic to the **_service_** - **_create_** method.